### DISCUSSION TITLE
Recommended APIs for events in Turbo Modules

### DISCUSSION THREAD

**Discussion by joemun** [reactwg/react-native-new-architecture#144] [Q&A]
👋 I'm having trouble finding official guidance/examples on the migration path for event emitters from Native Modules to Turbo Modules. Per https://reactnative.dev/docs/native-modules-android#sending-events-to-javascript, do Turbo Modules still support adding JS event listeners using `NativeEventEmitter`?

Is it possible to define listeners in the JS spec?  And for Android, what should the implementation look like? (ie. still create functions for `@ReactMethod addListener` and `@ReactMethod removeListeners`?)

Thanks!

**Comment by flokol120** [comment]
@joemun Maybe take a look here: https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/swift-event-emitter (If you did not know it existed :) )

It's for iOS only unfortunately, but it states:

> Note: The addListener and removeListeners implementations will be provided by React Native.

But I've already noticed, that the Android Spec Interface forces you to implement `addListener ` as well as `removeListeners `. However you can just leave the implementations empty. At least it worked in my use case. But I'd also like to know if this is intentional :D.

**Reply by joemun** [reply]
Thanks for responding. Yes, I'm primarily interested in the Android side.

Looks like event emitting still works when passing a turbo module into a `NativeEventEmitter`, and using its `addListener` API.  But I'm curious about whether this is the long-term API that uses the new JSI when New Arch is enabled, or is this only working because the legacy bridge is still in place.

**Comment by cortinico** [comment]
The recommended API to dispatch events in the New Architecture (and also in the old one) on Android is the Event Dispatcher. You can find an example of its usage here: https://github.com/reactwg/react-native-new-architecture/discussions/137#discussioncomment-6281398

**Reply by joemun** [reply]
Thank you for responding! but I think there was a misunderstanding in the question (apologies if my post wasn't clear).

The question wasn't about how to dispatch events from native to JS, but instead, the inverse.  Is there a type-safe way for JS to define and listen for events in Turbo Modules?  With Fabric Components, you can define Event Handlers in the JS spec.  However, with Turbo Modules I couldn't get that to work. Instead, you have to perform a cast from the turbo module interface to `NativeModule` when using `NativeEventEmitter`. Also the event handlers can't be defined in the JS spec.

For example:
```
const TurboModuleEmitter = new NativeEventEmitter(SampleTurboModule as unknown as NativeModule);
TurboModuleEmitter.addListener('eventName', ...);
```

Hope it's clear this time, thanks.

**Reply by DanielFrTB** [reply]
@cortinico, Does Event Dispatcher also recommended for iOS on the new arch?

**Reply by cortinico** [reply]
@DanielFrTB, nope. `EventDispatcher` is a class that exists only in Android for historical reasons.

**Reply by DanielFrTB** [reply]
@cortinico - so what is the solution for it on iOS new arch ?

**Reply by cortinico** [reply]
On iOS, you should use the Event Emitters that are generated by codegen

**Reply by DanielFrTB** [reply]
@cortinico Is there any example for using events on iOS using TurboModules  ?

**Reply by cipolleschi** [reply]
@DanielFrTB have a look at this doc: https://github.com/reactwg/react-native-new-architecture/blob/main/docs/turbo-modules.md#add-event-emitting-capabilities

This is for 0.76 and typed EventEmitters. Untyped event emitters should also be available on previous versions on React Native (<0.76) but we forgot to expose the TS type for the EventEmitter, so you'll probably see a type not found error, but it should actually work.

**Reply by DanielFrTB** [reply]
Thanks @cipolleschi

**Reply by abumubaarak** [reply]
@cipolleschi can you perhaps give an example for (<0.76)?

Thanks

**Reply by abumubaarak** [reply]
@cipolleschi Following the example at [test_fixtures](https://github.com/facebook/react-native/blob/main/packages/react-native-codegen/src/parsers/typescript/modules/__test_fixtures__/fixtures.js) for NATIVE_MODULE_WITH_EVENT_EMITTERS does not generate anything.

**Reply by Gregoirevda** [reply]
@cipolleschi is this still the case in recent versions of react-native with event emitters being code generated? eg
```ts
readonly onKeyAdded: CodegenTypes.EventEmitter<KeyValuePair>;
```
I see libraries using `new NativeEventEmitter(NativeModules.Module);` instead of what is explained in the [readme](https://reactnative.dev/docs/next/the-new-architecture/native-modules-custom-events?platforms=android#step-3-update-the-app-code). Is that fine?

**Reply by cipolleschi** [reply]
I think that the `NativeEventEmitter` is a deprected, legacy arch way to do it. We are keeping this around for the interop layers, but the fact that it accesses the `NativeModules.Module` it is clearly a sign that it is from Legacy Arch and should not be used for TM.

**Reply by tien** [reply]
@cipolleschi the new codegen EventEmitter doesn't seem to have a way to track if there are any active listeners like the old NativeEventEmitter? Which is useful if you need to spin up and then cleanup some resources.
